synapse/storage/databases/main/relations.py

```diff
@@ -978,6 +984,59 @@ def _get_thread_id(txn: LoggingTransaction) -> str:
 
         return await self.db_pool.runInteraction("get_thread_id", _get_thread_id)
 
+    @cached()
```

Comment:
Is this safe to cache? Sanity checks:

- is it possible for a given event to belong to more than one thread?
- is it possible that an event could belong to some thread at one moment in time, then later not belong to a thread (e.g. due to state resolution invalidating something that was previously accepted)

Reply:
> * is it possible for a given event to belong to more than one thread?

No.

> * is it possible that an event could belong to some thread at one moment in time, then later not belong to a thread (e.g. due to state resolution invalidating something that was previously accepted)

Only if the event is redacted, which is handled via invalidation in this PR.

Reply:
Thanks. The key thing I hadn't realised is that an event can only cite at most one other event that it relates to. This makes the relationships graph a tree.